docs/source/concepts/InstrumentAccessLayers.rst

```diff
@@ -31,216 +33,170 @@ ____________
-* ``l2(wsIndex)``
-* ``twoTheta(wsIndex)``
-* ``signedTwoTheta(wsIndex)``
-* ``position(wsIndex)``
+* ``isDetector(componentIndex)``
+* ``detectorsInSubtree(componentIndex)``
+* ``componentsInSubtree(componentIndex)``
+* ``position(componentIndex)``
+* ``rotation(componentIndex)``
+* ``hasParent(componentIndex)``
+* ``parent(componentIndex)``
+* ``position(componentIndex)``
+* ``solidAngle(componentIndex)``
+* ``scaleFactor(componentIndex)``
+* ``sourcePosition()``
+* ``samplePosition()``
+* ``l1()``
 
-The try/catch pattern for checking if a spectrum has a detector might look something like the following before/after example.
+The following methods are useful helpers on ``ComponentInfo`` that allow the extraction of the **component index** for key components
 
-**Before refactoring**
-
-.. code-block:: c++
+* ``root()``
+* ``source()``
+* ``sample()``
 
-  try {
-    Geometry::IDetector_const_sptr det = ws.getDetector(wsIndex);
-    // do stuff
-  catch (Exception::NotFoundError &) {
-    // do exception
-    return false;
-  }
-
-**After - check it has some detectors**
-
-.. code-block:: c++
+**Indexing**
 
-  #include "MantidAPI/SpectrumInfo.h"
+The ``ComponentInfo`` object is accessed by an index going from 0 to the number of components (including the instrument iteself). **ALL Detector Indexes have a Component Index which is the EXACT same representation**, this is an important point to understand. In other words, a detector with a Detector Index of 5, for the purposes of working with a ``DetectorInfo`` and  will have a Component Index of 5, when working with a ``ComponentInfo``. Explained in yet another way: The first 0 - n components referenced in the ``ComponentInfo`` are detectors, where n is the total number of detectors. This gurantee can be leveraged to provide speedups, as some of the examples will show.  
 
-  ...
+A ``ComponentID`` for compatiblity with older code, and be extracted from ``ComponentInfo::componentID(componentIndex)``, but such calls should be avoided where possible.
 
-  const auto &spectrumInfo = ws->spectrumInfo();
+It is also possible to use the method ``componentInfo.indexOf(componentID)`` to get the index for a particular component ID. However, this is a call to a lookup in an unordered map, so is an expensive calculation which should be avoided where possible.
 
-  if (spectrumInfo.hasDetectors(wsIndex)) {
-    // do stuff
-  } else {
-    // do exception
-    return false;
-  }
+**One should NEVER expose a Component Index or Detector Index through a public facing (python, gui, ...) to an end user**. Detector Index and Component Indexes are internal concepts for fast enumeration. It is however desirable to translate from a ``ComponentIndex`` via ``ComponentInfo::indexOf`` to as early as possible and in such a way to avoid repeated calls to this method, as stated above. Likewise, conversion back to a ``ComponentIndex``, if so required, should be done as infrequently and, as late as possible.
```

Comment:
Well, it will be on the Python interface. So maybe write "public facing (algorithm property, gui, ...)"?